tests/README.md

```diff
@@ -0,0 +1,1510 @@
+
+```
+ilModule{Module name}Suite.php
+```
+
+For service test suites the following pattern is applied:
+
+```
+ilService{Service name}Suite.php
+```
+
+<a name="new-parts-src"></a>
+#### New parts (/src)
+The test source for the new parts of ILIAS are located in the *tests* directory,
+which is located in the web root directory of ILIAS. The structure should be the same
+than the *src* directory where the actual implementation lives.
+
+```
+<webroot>/
+|
+-->  src
+|     |
+|     ------> <your service>/
+|     |             |
+|     |             --------> <your folders structur>/
+|     |                                |
+|     |                                ------> <your class>
+|     |
+|     ------> <other services>/
+|
+--> tests
+      |
+      ------> <your service>/
+      |             |
+      |             --------> <your folders structur>/
+      |                                |
+      |                                ------> <your test class>
+      |
+      ------> <other services>/
+```
+
+<a name="collector-ilglobalsuite-1"></a>
+##### Collector ilGlobalSuite
+The new tests are loaded from the *tests* directory which is located at the ILIAS web root. Each class which extends the *PHPUnit_Framework_TestCase* is loaded as test class.
+
+The new collector part must be updated with the PHPUnit 6 migration because the *TestCase* class no longer extends the legacy class *PHPUnit_Framework_TestCase*.
+The legacy class *PHPUnit_Framework_TestCase* will be entirely removed with PHPUnit 6.
+
+<a name="unit-test-structure"></a>
+### Unit-Test structure
+Each unit test is usually structured into three parts: arrange, act and assert. This are also known as
+the triple-A (see Langr 2015).
+
+- **Arrange** A proper system state is created by creating objects and interacting with them.
+- **Act** Invoke the part of the code which should be tested. This is usually one method call.
+- **Assert** Verify that the executed code behaves as expected. For example the verification of a return value
+or state of any objects involved. It can also involve verifications of interactions between objects with the
+help of mocks.
+
+If there is the need to clean up resources, a fourth step should be added.
+- **After** Ensures the cleanup of the used resources.
+
+All parts should be visually separated by a blank line to highlight the different parts.
+
+Example ILIAS Filesystem service (LegacyPathHelperTest):  
+
+```php
+<?php
+	/**
+	 * @Test
+	 * @small
+	 */
+	public function testCreateRelativePathWithWebTargetWhichShouldSucceed() {
+		$expectedPath = 'testtarget/subdir';
+		$target = $this->webPath . '/' . $expectedPath;
+
+		$result = LegacyPathHelper::createRelativePath($target);
+
+		$this->assertEquals($expectedPath, $result);
+	}
+```
+
+<a name="good-tests-are-first"></a>
+### Good tests are FIRST
+Many problems while unit testing can be avoided by following the FIRST principles (see Langr 2015).
```

Comment:
Yes FIRST helps while writing unit tests. This is especially hard todo in the older parts of ILIAS, but for the new services FIRST should help to provide a solid set of unit tests.

The testability issues are of course not fixed by FIRST but I tried to give some hints with the SOLID chapter at the end of the document.

Reply:
My point is: name (some) of the problems that this principle could or will solves for us (i.e. ILIAS community) or me (as an individual dev) instead of just stipulating that general unit test problems can be avoided by the principles. This will help people to adopt and understand that principle.